source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,263 @@
+  // the implementation may have a fixed order or a nondeterministic order,
+  // but conceptually a map field is supposed to be order agnostic.
+  //
+  // We should definitely *not* sort repeated fields in general; maps, however,
+  // are represented on the wire and in reflection format as repeated fields.
+  // Since the order is canonically not part of the data in the case of maps,
+  // this means for consistent hashing it must be ensured that the fields are
+  // handled in a deterministic order.
+  //
+  // It may be that if we didn't sort them the order would be deterministic, but
+  // that would be a non-guaranteed implementation detail, and a future version
+  // of protobuf could potentially change the internal map representation. Sorting
+  // also means theoretically we could produce a cross-language compatible hash;
+  // golang, for example, explicitly randomly orders map fields in its implementation.
+  ASSERT(field->is_map());
+  auto sorted_entries =
+      sortedMapField(reflection->GetRepeatedFieldRef<Protobuf::Message>(message, field));
+  const Protobuf::Descriptor* map_descriptor = sorted_entries.begin()->get().GetDescriptor();
+  const Protobuf::FieldDescriptor* key_field = map_descriptor->map_key();
+  const Protobuf::FieldDescriptor* value_field = map_descriptor->map_value();
+  for (const Protobuf::Message& entry : sorted_entries) {
+    seed = reflectionHashField(entry, key_field, seed);
+    seed = reflectionHashField(entry, value_field, seed);
+  }
+  return seed;
+}
+
+uint64_t reflectionHashField(const Protobuf::Message& message,
+                             const Protobuf::FieldDescriptor* field, uint64_t seed) {
+  using Protobuf::FieldDescriptor;
+  const Protobuf::Reflection* reflection = message.GetReflection();
+  {
+    const int q = field->number();
+    seed =
+        HashUtil::xxHash64(absl::string_view{reinterpret_cast<const char*>(&q), sizeof(q)}, seed);
+  }
+  switch (field->cpp_type()) {
+  case FieldDescriptor::CPPTYPE_INT32:
+    REFLECTION_FOR_EACH(Int32, int32_t);
+    break;
+  case FieldDescriptor::CPPTYPE_UINT32:
+    REFLECTION_FOR_EACH(UInt32, uint32_t);
+    break;
+  case FieldDescriptor::CPPTYPE_INT64:
+    REFLECTION_FOR_EACH(Int64, int64_t);
+    break;
+  case FieldDescriptor::CPPTYPE_UINT64:
+    REFLECTION_FOR_EACH(UInt64, uint64_t);
+    break;
+  case FieldDescriptor::CPPTYPE_DOUBLE:
+    REFLECTION_FOR_EACH(Double, double);
+    break;
+  case FieldDescriptor::CPPTYPE_FLOAT:
+    REFLECTION_FOR_EACH(Float, float);
+    break;
+  case FieldDescriptor::CPPTYPE_BOOL:
+    REFLECTION_FOR_EACH(Bool, bool);
+    break;
+  case FieldDescriptor::CPPTYPE_ENUM:
+    if (field->is_repeated()) {
+      int c = reflection->FieldSize(message, field);
+      for (int i = 0; i < c; i++) {
+        int v = reflection->GetRepeatedEnumValue(message, field, i);
+        seed = HashUtil::xxHash64(absl::string_view{reinterpret_cast<char*>(&v), sizeof(v)}, seed);
```

Comment:
I guess we don't care whether we get the same hash values on little-endian machines vs big-endian machines; this is only used for building up in-memory tables?

Reply:
It would be nice not to have to hash numbers using the bytes as a string-view. But I couldn't find an easy way to use xxhash to combine a hash from a number with other things.

Absl hashers make that easier but will give you different results on every process invocation which seems bad for the file cache.

Reply:
Working on it: https://github.com/envoyproxy/envoy/pull/31084